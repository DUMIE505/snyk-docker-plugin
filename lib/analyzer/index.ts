import * as osReleaseDetector from './os-release-detector';
import * as imageInspector from './image-inspector';
import * as apkAnalyzer from './apk-analyzer';
import * as aptAnalyzer from './apt-analyzer';
import * as rpmAnalyzer from './rpm-analyzer';
import * as binariesAnalyzer from './binaries-analyzer';
const debug = require('debug')('snyk');

export {
  analyze,
};

async function analyze(targetImage: string, options?: any) {
  const [
    imageInspection,
    osRelease,
    results,
  ] = await Promise.all([
    imageInspector.detect(targetImage, options),
    osReleaseDetector.detect(targetImage, options),
    Promise.all([
      apkAnalyzer.analyze(targetImage, options),
      aptAnalyzer.analyze(targetImage, options),
      rpmAnalyzer.analyze(targetImage, options),
    ]).catch((err) => {
      debug(`Error while running analyzer: '${err}'`);
      throw new Error(`Failed to detect installed OS packages: '${err}'`);
    }),

  ]);

  const { installedPackages, pkgManager } =
    getInstalledPackages(results as any[]);
  let binaries;
  try {
    binaries = await binariesAnalyzer.analyze(
      targetImage, installedPackages, pkgManager, options);
  } catch (err) {
    debug(`Error while running binaries analyzer: '${err}'`);
    throw new Error('Failed to detect binaries versions');
  }

  return {
    imageId: imageInspection.Id,
    osRelease,
    results,
    binaries,
    imageLayers: imageInspection.RootFS && imageInspection.RootFS.Layers,
  };
}

function getInstalledPackages(results: any[]):
 {installedPackages: string[], pkgManager?: string} {
  const dockerAnalysis = results.find((res) => {
    return res.Analysis && res.Analysis.length > 0;
  });

  if (!dockerAnalysis) {
    return { installedPackages: [] };
  }
  const installedPackages = dockerAnalysis.Analysis.map((pkg) => pkg.Name);
  let pkgManager = dockerAnalysis.AnalyzeType;
  if (pkgManager) {
    pkgManager = pkgManager.toLowerCase();
  }
  return { installedPackages, pkgManager };
}
