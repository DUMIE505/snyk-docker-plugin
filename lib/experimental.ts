import { PluginResponse } from "./types";

export async function experimentalAnalysis(
  options: any,
): Promise<PluginResponse> {
  // assume Distroless scanning
  return distroless(options);
}

// experimental flow expected to be merged with the static analysis when ready
export async function distroless(options: any): Promise<PluginResponse> {
  // assumption #1: the image is present in the local Docker daemon
  // assumption #2: the `docker` binary is available locally
  throw new Error("not implemented");
}
