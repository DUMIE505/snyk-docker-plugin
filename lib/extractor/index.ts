import { extractDockerArchiveLayers } from "./layer";
import { ExtractAction, ExtractedLayers } from "./types";

/**
 * Given a path on the file system to a docker-archive, open it up to inspect the layers
 * and look for specific files. File content can be transformed with a custom callback function if needed.
 * @param fileSystemPath Path to an existing docker-archive.
 * @param extractActions This denotes a file pattern to look for and how to transform the file if it is found.
 * By default the file is returned raw if no processing is desired.
 */
async function getDockerArchiveLayers(
  fileSystemPath: string,
  extractActions: ExtractAction[],
): Promise<ExtractedLayers> {
  const layers = await extractDockerArchiveLayers(
    fileSystemPath,
    extractActions,
  );

  if (!layers) {
    return {};
  }

  const result: ExtractedLayers = {};

  // TODO: This removes the information about the layer name, maybe we would need it in the future?

  for (const layer of layers) {
    // go over extracted files products found in this layer
    for (const filename of Object.keys(layer)) {
      // file was not found
      if (!Reflect.has(result, filename)) {
        result[filename] = layer[filename];
      }
    }
  }

  return result;
}

export { extractDockerArchiveLayers, getDockerArchiveLayers };
