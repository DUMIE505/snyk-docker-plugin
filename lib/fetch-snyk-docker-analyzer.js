const os = require('os');
const fs = require('fs');
const fsExtra = require('fs-extra');
const path = require('path');
const request = require('request');
const ciInfo = require('ci-info');
const ProgressBar = require('progress');
const pkgInfo = require('../package.json')

module.exports = fetch

const version = pkgInfo['snyk-docker-analyzer']['version'];

function getBinaryName() {
  const arch = os.arch();
  if (arch !== 'x64') {
    throw new Error(`Unsupported arch ${arch} - only amd64 is supported`);
  }
  const suffix = '';
  var platform = os.platform();
  if (platform === 'win32') {
    platform = 'windows';
    suffix = '.exe';
  }
  return `snyk-docker-analyzer-${platform}-amd64${suffix}`;
}

function getBinaryLocalPath() {
  const name = getBinaryName();
  return path.join(__dirname, '../bin/', version, name)
}

function fetch() {
  return new Promise((resolve, reject) => {
    try {
      const localPath = getBinaryLocalPath()
      if (fs.existsSync(localPath)) {
        return resolve(localPath);
      }
      fsExtra.ensureDirSync(path.dirname(localPath));
      const downloadUrl =
        `https://s3.amazonaws.com/snyk-docker-analyzer-releases/${version}/${getBinaryName()}`; // jscs:ignore maximumLineLength

      var bar;
      const fsStream = fs.createWriteStream(localPath + '.part');
      const req = request(downloadUrl);
      req
        .on('response', function (res) {
          if (res.statusCode >= 400) {
            var err = new Error(
              'Bad HTTP response for snyk-docker-analyzer download');
            err.statusCode = res.statusCode;
            fsStream.destroy();
            reject(err);
            req.abort();
            return;
          }

          if (ciInfo.isCI) {
            console.log(`downloading ${getBinaryName()} ...`);
          } else {
            const total = parseInt(res.headers['content-length'], 10);
            bar = new ProgressBar(`downloading ${getBinaryName()} [:bar] :rate/Kbps :percent :etas remaining`, { // jscs:ignore maximumLineLength
              complete: '=',
              incomplete: '.',
              width: 20,
              total: total / 1000,
            });
          }
        })
        .on('data', function (chunk) {
          if (bar) {
            bar.tick(chunk.length / 1000);
          }
        })
        .on('error', function (err) {
          reject(err);
        })
        .pipe(fsStream)
        .on('error', function (err) {
          fsStream.destroy();
          reject(err);
        })
        .on('finish', function () {
          fs.renameSync(localPath + '.part', localPath);
          const CHMOD_WITH_EXEC = 0755;
          fs.chmodSync(localPath, CHMOD_WITH_EXEC);
          resolve(localPath);
        })
    } catch (err) {
      reject(err);
    }
  })
}
