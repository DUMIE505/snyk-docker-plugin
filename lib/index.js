var fs = require('fs');
var path = require('path');
var subProcess = require('./sub-process');
var fetchSnykDockerAnalyzer = require('./fetch-snyk-docker-analyzer');

module.exports = {
  inspect: inspect,
};

function inspect(root, targetFile) {
  const targetImage = root;
  return fetchSnykDockerAnalyzer()
  .then(function (analyzerBinaryPath) {
    return Promise.all([
      getMetaData(),
      getDependencies(analyzerBinaryPath, targetImage),
    ])
    .then(function (result) {
      return {
        plugin: result[0],
        package: result[1],
      };
    });
  })
}

function getMetaData() {
  return subProcess.execute('docker', ['version'])
  .then(function (output) {
    var runtime;
    versionMatch = /Version:\s+(.*)\n/.exec(output);
    if (versionMatch) {
      runtime = 'docker ' + versionMatch[1]
    }

    return {
      name: 'snyk-docker-plugin',
      runtime: runtime,
    };
  });
}

function getDependencies(analyzerBinaryPath, targetImage) {
  console.log(`analyzing ${targetImage} Docker image`);
  return subProcess.execute(
    analyzerBinaryPath,
    buildArgs(targetImage)
  )
  .then(function (output) {
    scanResults = JSON.parse(output);
    return Promise.resolve(convertDependecies(targetImage, scanResults));
  })
  .catch(function (error) {
    if (typeof error === 'string') {
      if (error.indexOf('command not found') !== -1) {
        throw new Error('Snyk docker CLI wasn\'t found')
      }
      errorMsg = error;
      errorMatch = /msg="(.*)"/g.exec(errorMsg)
      if (errorMatch) {
        errorMsg = errorMatch[1]
      }
      throw new Error(errorMsg);
    }

    throw error;
  })
}

function convertDependecies(targetImage, scanResults) {
  targetSplit = targetImage.split(':');
  imageName = targetSplit[0];
  imageVersion = targetSplit[1];
  imageVersion = (imageVersion ? imageVersion : 'lateset');

  root = {};
  root.name = imageName;
  root.version = imageVersion;
  root.from = [imageName + '@' + imageVersion];

  var scanResult = scanResults.filter(function (res) {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  var pkgType;
  switch (scanResult.AnalyzeType) {
    case 'Apt': {
      pkgType = 'deb';
      break;
    }
    default: {
      pkgType = scanResult.AnalyzeType.toLowerCase()
    }
  }
  root.packageFormatVersion = pkgType + ':0.0.1';

  pkgs = scanResult['Analysis']

  root.dependencies = pkgs.reduce(function (acc, pkg) {
    name = pkg['Name'];
    version = pkg['Version'];

    acc[name] = {
      name: name,
      version: version,
      dependencies: {},
      from: [
        root['from'][0],
        name + '@' + version,
      ],
    }
    return acc;
  }, {});

  return root;
}

function buildArgs(targetImage) {
  var args = ['analyze', targetImage];
  return args;
}

function pathToPosix(fpath) {
  var parts = fpath.split(path.sep);
  return parts.join(path.posix.sep);
}
