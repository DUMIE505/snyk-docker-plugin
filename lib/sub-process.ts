import * as childProcess from "child_process";

export {
  execute,
  executeAsStream,
  CmdOutput,
  StreamData,
  ExecuteAsStreamCallback,
};
interface CmdOutput {
  stdout: string;
  stderr: string;
}

interface StreamData {
  data?: Buffer | string;
  err?: Buffer | string;
  exitCode?: number;
}

type ExecuteAsStreamCallback = (sd: StreamData) => any;

function execute(
  command: string,
  args?: string[],
  options?,
): Promise<CmdOutput> {
  const spawnOptions: any = { shell: true };
  if (options && options.cwd) {
    spawnOptions.cwd = options.cwd;
  }

  return new Promise((resolve, reject) => {
    let stdout = "";
    let stderr = "";

    const proc = childProcess.spawn(command, args, spawnOptions);
    proc.stdout.on("data", (data) => {
      stdout = stdout + data;
    });
    proc.stderr.on("data", (data) => {
      stderr = stderr + data;
    });

    proc.on("close", (code) => {
      const output = { stdout, stderr };
      if (code !== 0) {
        return reject(output);
      }
      resolve(output);
    });
  }) as Promise<CmdOutput>;
}

function executeAsStream(
  command: string,
  cb: ExecuteAsStreamCallback,
  args?: string[],
): Promise<void> {
  return new Promise((resolve, reject) => {
    const spawnOptions: any = { shell: true };

    const proc = childProcess.spawn(command, args, spawnOptions);
    proc.stdout.on("data", (data) => {
      cb({ data });
    });

    proc.stderr.on("data", (data) => {
      cb({ err: data });
    });

    proc.on("close", (exitCode) => {
      cb({ exitCode });

      resolve();
    });
  });
}
