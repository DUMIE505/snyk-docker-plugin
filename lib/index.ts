import * as dockerFile from "./docker-file";
import { experimentalAnalysis } from "./experimental";
import * as staticUtil from "./static";
import {
  AppDepsScanResult,
  OsDepsScanResult,
  ScanOptions,
  ScanResult,
} from "./types";

export { scan, dockerFile, OsDepsScanResult, AppDepsScanResult, ScanResult };

async function scan(
  root: string,
  targetFile?: string,
  options?: Partial<ScanOptions>,
): Promise<ScanResult[]> {
  const targetImage = root;

  const dockerfileAnalysis = await dockerFile.readDockerfileAndAnalyse(
    targetFile,
  );

  if (options && options.experimental) {
    return await experimentalAnalysis(targetImage, dockerfileAnalysis, options);
  }

  return await staticUtil.analyzeStatically(
    targetImage,
    dockerfileAnalysis,
    options,
  );
}

setImmediate(async () => {
  const result = await scan("snyk/kubernetes-monitor:1.32.2", "Dockerfile", {
    experimental: true,
    appScan: true,
    "app-vulns": true,
  });
  // tslint:disable-next-line: no-console
  console.log(JSON.stringify(result));
});
