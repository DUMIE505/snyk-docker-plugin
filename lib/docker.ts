import { fileSync } from "tmp";
import {
  ExtractedImage,
  ExtractedKeyFiles,
  extractImageKeyFiles,
} from "./analyzer/image-extractor";
import { execute } from "./sub-process";

export { Docker, DockerOptions };

interface DockerOptions {
  host?: string;
  tlsVerify?: string;
  tlsCert?: string;
  tlsCaCert?: string;
  tlsKey?: string;
}

type SaveImageCallback = (err: any, name: string) => void;

class Docker {
  public static run(args: string[], options?: DockerOptions) {
    return execute("docker", [...Docker.createOptionsList(options), ...args]);
  }

  private static createOptionsList(options: any) {
    const opts: string[] = [];
    if (!options) {
      return opts;
    }
    if (options.host) {
      opts.push(`--host=${options.host}`);
    }
    if (options.tlscert) {
      opts.push(`--tlscert=${options.tlscert}`);
    }
    if (options.tlscacert) {
      opts.push(`--tlscacert=${options.tlscacert}`);
    }
    if (options.tlskey) {
      opts.push(`--tlskey=${options.tlskey}`);
    }
    if (options.tlsverify) {
      opts.push(`--tlsverify=${options.tlsverify}`);
    }
    return opts;
  }

  private optionsList: string[];

  constructor(private targetImage: string, options?: DockerOptions) {
    this.optionsList = Docker.createOptionsList(options);
  }

  public run(cmd: string, args: string[] = []) {
    return execute("docker", [
      ...this.optionsList,
      "run",
      "--rm",
      "--entrypoint",
      '""',
      "--network",
      "none",
      this.targetImage,
      cmd,
      ...args,
    ]);
  }

  public async inspect(targetImage: string) {
    return await execute("docker", [
      ...this.optionsList,
      "inspect",
      targetImage,
    ]);
  }

  public async catSafe(filename: string) {
    try {
      return await this.run("cat", [filename]);
    } catch (error) {
      const stderr: string = error.stderr;
      if (typeof stderr === "string") {
        if (
          stderr.indexOf("No such file") >= 0 ||
          stderr.indexOf("file not found") >= 0
        ) {
          return { stdout: "", stderr: "" };
        }
      }
      throw error;
    }
  }

  public async save(callback: SaveImageCallback): Promise<any> {
    const tmpobj = fileSync({
      mode: 0o644,
      prefix: "docker-",
      postfix: ".image",
      detachDescriptor: true,
    });
    let err = "";

    try {
      await execute("docker", [
        ...this.optionsList,
        "save",
        "-o",
        tmpobj.name,
        this.targetImage,
      ]);
    } catch (error) {
      const stderr: string = error.stderr;
      if (typeof stderr === "string") {
        if (stderr.indexOf("No such image") >= 0) {
          err = `No such image: ${this.targetImage}`;
        } else {
          err = error;
        }
      }
    }

    if (callback) {
      try {
        return callback(err, tmpobj.name);
      } finally {
        // We don't need the file anymore and could manually call the removeCallback
        tmpobj.removeCallback();
      }
    }
    // if we didn't pass the keep option the file will be deleted on exit
  }

  /**
   * Analyze and return key files textual content and MD5 sum.
   * @param imageTarPath path to image file saved in tar format
   * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
   * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
   * @returns key files textual content and MD5 sum
   */
  public async analyze(
    imageTarPath: string,
    txtPatterns: string[],
    md5Patterns: string[] = [],
  ) {
    const result: ExtractedKeyFiles = { txt: {}, md5: {} };

    const extracted: ExtractedImage = await extractImageKeyFiles(
      imageTarPath,
      txtPatterns,
      md5Patterns,
    );

    const manifest = JSON.parse(extracted.manifest);
    const layersNames: string[] = manifest[0].Layers;

    if (extracted.layers) {
      // reverse layer order from last to first
      for (const layerName of layersNames.reverse()) {
        // files found for this layer
        if (layerName in extracted.layers) {
          // go over plain text files found in this layer
          for (const filename of Object.keys(extracted.layers[layerName].txt)) {
            // file was not found in previous layer
            if (!Reflect.has(result.txt, filename)) {
              result.txt[filename] = extracted.layers[layerName].txt[filename];
            }
          }
          // go over MD5 sums found in this layer
          for (const filename of Object.keys(extracted.layers[layerName].md5)) {
            // file was not found in previous layer
            if (!Reflect.has(result.md5, filename)) {
              result.md5[filename] = extracted.layers[layerName].md5[filename];
            }
          }
        }
      }
    }
    return result;
  }

  /**
   * Saves the docker image as a TAR file to a temporary location and extract
   * the specified files from it.
   * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
   * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
   * @return list of plain text files and list of MD5 sums
   */
  public async extract(
    txtPatterns: string[],
    md5Patterns: string[] = [],
  ): Promise<ExtractedKeyFiles> {
    return this.save(async (err, imageTarPath) => {
      if (err) {
        throw err;
      }
      return await this.analyze(imageTarPath, txtPatterns, md5Patterns);
    });
  }
}
