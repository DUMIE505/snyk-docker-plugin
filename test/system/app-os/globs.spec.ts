import { scan } from "../../../lib/index";
import { execute } from "../../../lib/sub-process";

describe("find globs tests", () => {
  afterAll(async () => {
    await execute("docker", ["image", "rm", "debian:10"]).catch();
  });

  it("should correctly return manifest files when detected by globs", async () => {
    const image = "debian:10";
    const pluginResult = await scan({
      path: image,
      globsToFind: {
        include: ["**/os-release"],
        exclude: [],
      },
    });

    expect(pluginResult).toMatchSnapshot();
  });
});
