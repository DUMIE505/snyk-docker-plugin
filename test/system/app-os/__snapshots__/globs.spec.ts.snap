// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find globs tests should correctly return manifest files when detected by globs 1`] = `
Array [
  Object {
    "contents": Object {
      "data": Array [
        80,
        82,
        69,
        84,
        84,
        89,
        95,
        78,
        65,
        77,
        69,
        61,
        34,
        68,
        101,
        98,
        105,
        97,
        110,
        32,
        71,
        78,
        85,
        47,
        76,
        105,
        110,
        117,
        120,
        32,
        49,
        48,
        32,
        40,
        98,
        117,
        115,
        116,
        101,
        114,
        41,
        34,
        10,
        78,
        65,
        77,
        69,
        61,
        34,
        68,
        101,
        98,
        105,
        97,
        110,
        32,
        71,
        78,
        85,
        47,
        76,
        105,
        110,
        117,
        120,
        34,
        10,
        86,
        69,
        82,
        83,
        73,
        79,
        78,
        95,
        73,
        68,
        61,
        34,
        49,
        48,
        34,
        10,
        86,
        69,
        82,
        83,
        73,
        79,
        78,
        61,
        34,
        49,
        48,
        32,
        40,
        98,
        117,
        115,
        116,
        101,
        114,
        41,
        34,
        10,
        86,
        69,
        82,
        83,
        73,
        79,
        78,
        95,
        67,
        79,
        68,
        69,
        78,
        65,
        77,
        69,
        61,
        98,
        117,
        115,
        116,
        101,
        114,
        10,
        73,
        68,
        61,
        100,
        101,
        98,
        105,
        97,
        110,
        10,
        72,
        79,
        77,
        69,
        95,
        85,
        82,
        76,
        61,
        34,
        104,
        116,
        116,
        112,
        115,
        58,
        47,
        47,
        119,
        119,
        119,
        46,
        100,
        101,
        98,
        105,
        97,
        110,
        46,
        111,
        114,
        103,
        47,
        34,
        10,
        83,
        85,
        80,
        80,
        79,
        82,
        84,
        95,
        85,
        82,
        76,
        61,
        34,
        104,
        116,
        116,
        112,
        115,
        58,
        47,
        47,
        119,
        119,
        119,
        46,
        100,
        101,
        98,
        105,
        97,
        110,
        46,
        111,
        114,
        103,
        47,
        115,
        117,
        112,
        112,
        111,
        114,
        116,
        34,
        10,
        66,
        85,
        71,
        95,
        82,
        69,
        80,
        79,
        82,
        84,
        95,
        85,
        82,
        76,
        61,
        34,
        104,
        116,
        116,
        112,
        115,
        58,
        47,
        47,
        98,
        117,
        103,
        115,
        46,
        100,
        101,
        98,
        105,
        97,
        110,
        46,
        111,
        114,
        103,
        47,
        34,
        10,
      ],
      "type": "Buffer",
    },
    "name": "os-release",
    "path": "/usr/lib",
  },
]
`;
